### PR TITLE
[feature](Cloud) Add S3 storage vault implementation

### PR DESCRIPTION
**Pull request by ByteYue**
## Proposed changes

Issue Number: close #xxx

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: cloud/src/meta-service/meta_service_resource.cpp [hunk 1]
```diff
@@ -404,6 +404,82 @@ static int add_hdfs_storage_vault(InstanceInfoPB& instance, Transaction* txn,
     return 0;
 }
 
+static std::optional<ObjectStoreInfoPB> create_object_info_with_encrypt(
+        InstanceInfoPB& instance, const ObjectStoreInfoPB& obj, bool sse_enabled,
+        MetaServiceCode& code, std::string& msg) {
```
Comment (platoneko) [inline cloud/src/meta-service/meta_service_resource.cpp:409]:
```suggestion
static void create_object_info_with_encrypt(
        const InstanceInfoPB& instance, const ObjectStoreInfoPB& obj, bool sse_enabled,
        ObjectStoreInfoPB& out, MetaServiceCode& code, std::string& msg) {
```

**Review by platoneko — CHANGES_REQUESTED** [review]

**Comment by ByteYue** [comment]
run buildall

**Comment by ByteYue** [comment]
run buildall

**Comment by ByteYue** [comment]
run buildall

**Comment by ByteYue** [comment]
run buildall

**Review by gavinchou — APPROVED** [review]

File: fe/fe-core/src/main/java/org/apache/doris/catalog/StorageVaultMgr.java [hunk 2]
```diff
@@ -159,4 +162,31 @@ private void alterSyncVaultTask() {
...
+
+    public void createS3Vault(StorageVault vault) throws DdlException {
+        S3StorageVault s3StorageVault = (S3StorageVault) vault;
+        Cloud.AlterObjStoreInfoRequest.Builder requestBuilder
+                = Cloud.AlterObjStoreInfoRequest.newBuilder();
+        requestBuilder.setOp(Cloud.AlterObjStoreInfoRequest.Operation.ADD_S3_VAULT);
+        Cloud.ObjectStoreInfoPB.Builder objBuilder = S3Properties.getObjStoreInfoPB(vault.getCopiedProperties());
+        Cloud.StorageVaultPB.Builder alterObjVaultBuilder = Cloud.StorageVaultPB.newBuilder();
+        alterObjVaultBuilder.setName(s3StorageVault.getName());
+        alterObjVaultBuilder.setS3Obj(objBuilder.build());
```
Comment (gavinchou) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/StorageVaultMgr.java:174]:
```suggestion
        alterObjVaultBuilder.setObjInfo(objBuilder.build());
```

File: fe/fe-core/src/main/java/org/apache/doris/catalog/StorageVault.java [hunk 3]
```diff
@@ -166,28 +174,23 @@ protected void replaceIfEffectiveValue(Map<String, String> properties, String ke
...
-        Cloud.HdfsVaultInfo.Builder builder = Cloud.HdfsVaultInfo.newBuilder();
-        builder.mergeFrom(vault.getHdfsInfo());
         TextFormat.Printer printer = TextFormat.printer();
-        row.add(printer.shortDebugString(builder));
+        if (vault.hasHdfsInfo()) {
+            Cloud.HdfsVaultInfo.Builder builder = Cloud.HdfsVaultInfo.newBuilder();
+            builder.mergeFrom(vault.getHdfsInfo());
+            row.add(printer.shortDebugString(builder));
+        }
+        if (vault.hasS3Obj()) {
```
Comment (gavinchou) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/StorageVault.java:187]:
```suggestion
        if (vault.hasObjInfo()) {
```

File: fe/fe-core/src/main/java/org/apache/doris/catalog/StorageVault.java [hunk 4]
```diff
@@ -166,28 +174,23 @@ protected void replaceIfEffectiveValue(Map<String, String> properties, String ke
...
+            Cloud.ObjectStoreInfoPB.Builder builder = Cloud.ObjectStoreInfoPB.newBuilder();
+            builder.mergeFrom(vault.getS3Obj());
```
Comment (gavinchou) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/StorageVault.java:189]:
```suggestion
            builder.mergeFrom(vault.getObjInfo());
```

**Review by platoneko — DISMISSED** [review]

**Comment by ByteYue** [comment]
run buildall

**Review by platoneko — APPROVED** [review]

**Pull request merged**